pkg/gce-pd-csi-driver/controller.go

```diff
@@ -517,37 +505,57 @@ func diskIsAttachedAndCompatible(volume *compute.Disk, instance *compute.Instanc
 	return false, nil
 }
 
-func pickTopology(top *csi.TopologyRequirement) (string, error) {
-	reqTop := top.GetRequisite()
-	prefTop := top.GetPreferred()
+func pickZonesFromTopology(top *csi.TopologyRequirement, numZones int) ([]string, error) {
+	reqZones, err := getZonesFromTopology(top.GetRequisite())
+	if err != nil {
+		return nil, fmt.Errorf("could not get zones from requisite topology: %v", err)
+	}
+	prefZones, err := getZonesFromTopology(top.GetPreferred())
+	if err != nil {
+		return nil, fmt.Errorf("could not get zones from preferred topology: %v", err)
+	}
 
-	// Pick the preferred topology in order
-	if len(prefTop) != 0 {
-		if prefTop[0].GetSegments() == nil {
-			return "", fmt.Errorf("preferred topologies specified but no segments")
+	if numZones <= len(prefZones) {
+		return prefZones[0:numZones], nil
+	} else {
+		zones := sets.String{}
+		// Add all preferred zones into zones
+		zones.Insert(prefZones...)
+		remainingNumZones := numZones - len(prefZones)
+		// Take all of the remaining zones from requisite zones
+		reqSet := sets.NewString(reqZones...)
+		prefSet := sets.NewString(prefZones...)
+		remainingZones := reqSet.Difference(prefSet)
+
+		if remainingZones.Len() < remainingNumZones {
+			return nil, fmt.Errorf("need %v zones from topology, only got %v unique zones", numZones, reqSet.Union(prefSet).Len())
 		}
-
-		// GCE PD cloud provider Create has no restrictions so just create in top preferred zone
-		zone, err := getZoneFromSegment(prefTop[0].GetSegments())
+		// Add the remaining number of zones into the set
+		// TODO: Maybe we should pick a random set to get
+		nSlice, err := pickRandNFromSlice(remainingZones.List(), remainingNumZones)
```

Comment:
I think we actually want the first zone to be random.  But the second zone should always be +1 from the first (sorted)